module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,235 @@
+// It processes events, transaction results, registers, collections, and transactions
+// from their respective in-memory storages to permanent storage.
+// Note: This method doesn't commit the batch - that should be done by the caller.
+func (p *Persister) AddToBatch(batch storage.Batch) error {
+	log := p.log.With().
+		Hex("block_id", logging.ID(p.executionResult.BlockID)).
+		Uint64("height", p.header.Height).
+		Logger()
+
+	log.Debug().Msg("adding execution data to batch")
+
+	start := time.Now()
+
+	if err := p.addEventsToBatch(batch); err != nil {
+		return err
+	}
+
+	if err := p.addResultsToBatch(batch); err != nil {
+		return err
+	}
+
+	if err := p.persistRegisters(); err != nil {
+		return err
+	}
+
+	if err := p.persistCollections(); err != nil {
+		return err
+	}
+
+	if err := p.persistTransactions(); err != nil {
+		return err
+	}
+
+	p.LastPersistedSealedExecutionResult = p.executionResult
+
+	duration := time.Since(start)
+	log.Debug().
+		Dur("duration_ms", duration).
+		Msg("successfully prepared execution data for persistence")
+
+	return nil
+}
+
+// addEventsToBatch adds events from in-memory storage to the batch.
+// If there are no events, this is a no-op.
+func (p *Persister) addEventsToBatch(batch storage.Batch) error {
+	if eventsList := p.inMemoryEvents.Data(); len(eventsList) > 0 {
+		if err := p.events.BatchStore(p.executionResult.BlockID, []flow.EventsList{eventsList}, batch); err != nil {
+			return fmt.Errorf("could not add events to batch: %w", err)
+		}
+		p.log.Debug().Int("event_count", len(eventsList)).Msg("added events to batch")
+	}
+
+	return nil
+}
+
+// addResultsToBatch adds transaction results from in-memory storage to the batch.
+// If there are no results, this is a no-op.
+func (p *Persister) addResultsToBatch(batch storage.Batch) error {
+	if results := p.inMemoryResults.Data(); len(results) > 0 {
+		if err := p.results.BatchStore(p.executionResult.BlockID, results, batch); err != nil {
+			return fmt.Errorf("could not add transaction results to batch: %w", err)
+		}
+		p.log.Debug().Int("result_count", len(results)).Msg("added transaction results to batch")
+	}
+
+	return nil
+}
+
+// persistRegisters persists registers from in-memory to permanent storage.
+// If there are no registers, this is a no-op.
+func (p *Persister) persistRegisters() error {
+	if registerData := p.inMemoryRegisters.Data(); len(registerData) > 0 {
+		start := time.Now()
+
+		if err := p.registers.Store(registerData, p.header.Height); err != nil {
+			return fmt.Errorf("could not persist registers: %w", err)
+		}
+
+		p.log.Debug().
+			Int("register_count", len(registerData)).
+			Dur("duration_ms", time.Since(start)).
+			Msg("persisted registers")
+	}
+
+	return nil
+}
+
+// persistCollections persists collections from in-memory to permanent storage.
+// It skips collections that already exist in permanent storage.
+// If there are no collections, this is a no-op.
+func (p *Persister) persistCollections() error {
```

Comment:
I think collections and transactions should also be added within the batch. We may need to add new `Batch*` methods for the storage objects if they don't exist.